api/docs/scatter_gather_emulation.dox

```diff
@@ -0,0 +1,230 @@
+
+Owner: [Hendrik Greving](https://github.com/hgreving2304)
+
+As described above, we can simplify work for DR clients by replacing the scatter and gather instructions with a functionally equivalent sequence of scalar stores and loads. The expanded sequence is the unrolled version of the following loop:
+```
+num_accesses = vector_size / element_size
+for i = 0, 1, ..., (num_accesses-1), do
+  extract mask for the ith access from mask reg or mask vector
+  if mask is set, then
+    extract ith element of index vector
+    compute address = base + ith index element
+    if instr_is_gather, then
+      load data from address into a scalar reg
+      insert scalar data into destination vector
+    else // instr_is_scatter
+      Extract scalar data from source vector to scalar reg
+      Store data from scalar reg to address
+    done
+    clear ith mask in mask reg or mask vector
+  done
+done
+```
+
+Due to the x86 ISA, the extraction/insertion of the scalar value from/to the vector may involve multiple steps, e.g. to extract a 32-bit scalar value from a 512-bit `zmm` reg, we first need to extract a 128-bit `xmm` from it.
+
+drmgr in DR provides multiple phases of instrumentation. This expansion is done in the first phase known as app2app. As the name suggests, this phase is intended to transform app instructions to equivalent instructions. For simplicity, we also separate out the scatter or gather instructions from its basic block and create a separate fragment with only the expanded sequence. The logic for expanding scatter and gather instructions is implemented in the drx extension library as [`drx_expand_scatter_gather`](https://github.com/DynamoRIO/dynamorio/blob/eb5d5af8e3444912c9f3f70e5ebf7969252ee4d6/ext/drx/drx.h#L538), and can be used by any client that needs it, including drcachesim. This support was added by [commit](https://github.com/DynamoRIO/dynamorio/commit/4359ef134e47942004c09db04b54593579763186).
+
+As an example, the following are the expansions of some instructions.
+
+Expansion for `vpgatherdd`: https://gist.github.com/abhinav92003/637c07bac785428550f05c18319f80c2
+
+Expansion for `vpscatterdd`: https://gist.github.com/abhinav92003/4a78ec4ad1621cbf6ca905a6774cb389
+
+We also need to make sure that if some load or store faults, we restore the application state properly. While the spilled registers obtained from drreg are restored by the drreg state restoration logic, drx still needs to restore the scratch mask registers that are spilled manually to a GPR and ensure that the bit for the faulting access is cleared too. In case a fault or some other state restore event like DR detach happens, we walk the expanded sequence using a state machine till we reach the fault pc, keeping track of the state that needs to be restored ([commit](https://github.com/DynamoRIO/dynamorio/commit/de511a72197e9b04a9732f33622fe7bc7fd12623), [commit](https://github.com/DynamoRIO/dynamorio/commit/e54913b4845794b96e8682ba6a214d54e10741a0)).
```

Comment:
The mask bits for the completed memrefs have to be cleared, but not the faulting one?

Reply:
Based on this comment
https://github.com/DynamoRIO/dynamorio/blob/faf502728968e8d003b912b6771b3df5ed9bc7aa/ext/drx/drx.c#L3701
I believe the faulting access's bit needs to be zeroed.

Reply:
But how do you re-execute the instruction then?  You would want to re-try the not-completed previously-faulted access, but if the bit is already zeroed it won't even try it on the re-execution.

Reply:
yeah, I'm confused by this too. I tried this out natively and I don't see the faulting mask bit being zeroed. I'm checking to see if I mis-understood that comment.

Reply:
Still looking into this. Pushed the other changes for review for now.

Reply:
Ah, figured this out. The code at the above comment and other similar comments in this file (e.g. the one at https://github.com/DynamoRIO/dynamorio/blob/faf502728968e8d003b912b6771b3df5ed9bc7aa/ext/drx/drx.c#L3174) applies only when there's a fault between L180 and L187 in the doc I just added (non-inclusive). This is when the load completed but there was a fault before we could clear the mask.

I don't think this can happen in any instruction in the expanded sequence itself (there's no memory instruction there), but maybe some other client's instrumentation or other async state restore event may trigger this path.

I'll fix the documentation to clarify this.

Reply:
Done.